File: 2023/1/solution.jl
f = open("input")

digits = ["1", "2", "3", "4", "5", "6", "7", "8", "9"]
sum1 = 0
for line in readlines(f)
    digdexfirst = minimum([isnothing(findfirst(d, line)) ? 10000 : findfirst(d, line)[1] for d in digits])
    digdexlast = maximum([isnothing(findlast(d, line)) ? -1 : findlast(d, line)[1] for d in digits])

    sum1 = sum1 + parse(Int64, string(line[digdexfirst], line[digdexlast]))
end

print(sum1)
print("\n")
close(f)

digits2 = ["1", "2", "3", "4", "5", "6", "7", "8", "9", 
         "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
numbers = cat(1:9, 1:9; dims = 1)
digits_dict = Dict([(digits2[i], numbers[i]) for i in 1:size(digits2)[1]])

f = open("input")

sum2 = 0
for line in readlines(f)
    digdexfirst = [findfirst(d, line) for d in digits2]
    digdexlast = [findlast(d, line) for d in digits2]

    dig1 = digits_dict[line[digdexfirst[argmin([isnothing(d) ? 10000 : d[1] for d in digdexfirst])]]] 
    dig2 = digits_dict[line[digdexlast[argmax([isnothing(d) ? -1 : d[1] for d in digdexlast])]]]

    sum2 = sum2 + 10 * dig1 + dig2
end

print(sum2)

close(f)

File: 2023/1/solution.R
input <- readLines("input")

sumnums <- function(inp) {
strsplit(inp, "[a-z]+") |> 
  sapply(\(x) {
    numstr <- paste0(x, collapse = "")
    paste0(substr(numstr, 1, 1), substr(numstr, nchar(numstr), nchar(numstr))) |> 
      as.numeric()
    }) |> 
  sum()
}

sumnums(input)

## part two

torep <- c("one", "two", "three", "four", 
           "five", "six", "seven", "eight", 
           "nine")

lookup <- c(1:9, 1:9)
names(lookup) <- c(torep, as.character(1:9))

first <- rep(1e8, length(input))
last <- rep(-1, length(input))
first.length <- last.length <- rep(NA, length(input))

for(j in seq_along(lookup)) {
  
  mat <- gregexpr(names(lookup[j]), input)
  thismin <- sapply(mat, \(m){
    m[m == -1] <- 1e8
    c(min(m), attr(m, "match.length")[which.min(m)]) 
  })
  first.length <- ifelse(thismin[1, ] < first, thismin[2, ], first.length)
  first <- pmin(first, thismin[1,], na.rm = TRUE)
  
  thismax <- sapply(mat, \(m){
    c(max(m), attr(m, "match.length")[which.max(m)]) 
  })
  last.length <- ifelse(thismax[1, ] > last, thismax[2, ], last.length)
  last <- pmax(last, thismax[1,], na.rm = TRUE)
  
}


as.numeric(paste0(lookup[substr(input, first, first + first.length - 1)], 
       lookup[substr(input, last, last + last.length - 1)])) |> sum()
